src/cpp/shared_core/system/PosixSystem.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * PosixSystem.cpp
+ *
+ * Copyright (C) 2009-19 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant to the terms of a commercial license agreement
+ * with RStudio, then this program is licensed to you under the following terms:
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ * Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#include <shared_core/system/PosixSystem.hpp>
+
+#include <csignal>
+#include <grp.h>
+#include <memory.h>
+#include <pwd.h>
+#include <sys/prctl.h>
+
+#include <shared_core/Error.hpp>
+#include <shared_core/system/User.hpp>
+
+namespace rstudio {
+namespace core {
+namespace system {
+namespace posix {
+
+Error enableCoreDumps()
+{
+#ifndef __APPLE__
+   int res = ::prctl(PR_SET_DUMPABLE, 1);
+   if (res == -1)
+      return systemError(errno, ERROR_LOCATION);
+#endif
+
+   return Success();
+}
+
+Error ignoreSignal(int in_signal)
+{
+   struct sigaction sa;
+   ::memset(&sa, 0, sizeof(sa));
+
+   sa.sa_handler = SIG_IGN;
+   int result = ::sigaction(in_signal, &sa, nullptr);
+   if (result != 0)
+   {
+      Error error = systemError(result, ERROR_LOCATION);
+      error.addProperty("signal", in_signal);
+      return error;
+   }
+
+   return Success();
+}
+
+bool realUserIsRoot()
+{
+   return ::getuid() == 0;
+}
+
+Error restoreRoot()
+{
+   // Reset error state.
+   errno = 0;
+
+   // Change the effective user to root.
+   if (::seteuid(0) < 0)
+      return systemError(errno, ERROR_LOCATION);
+   // Verify
+   if (::geteuid() != 0)
+      return systemError(EACCES, ERROR_LOCATION);
+
+   // Get user info to use in group calls
+   struct passwd* pPrivPasswd = ::getpwuid(0);
```

Comment:
IIUC errors here and below would lead us to still having root, just not having the effective group set. Is it acceptable for us to end up in this state, or should we try to make sure this function is 'atomic'? (not sure atomic is the best word but it feels like the closest analogue here)

Reply:
That's an interesting question. I think in the cases I've seen if this function returns an error the process exits. That being said, it would make the function more "safe", I think, which always seems like a good idea.

At any rate, this is the existing behaviour so I doubt it will cause issues if we don't change it.

Reply:
That makes sense. If failures here are handled as process exits then it's not worth changing anything here.

Reply:
Sure, I'll double check the usages of this function.